File: src/Vulkan/Swap_Chain.cpp
#include "Vulkan.h"

#include <cstdint>
#include <algorithm>
#include <iostream>

void querySwapChainSupport(VkPhysicalDevice device, Vulkan& vulkan) {
  vkGetPhysicalDeviceSurfaceCapabilitiesKHR(device, vulkan.surface, &vulkan.swapchainDetails.capabilities);

  uint32_t formatCount;
  vkGetPhysicalDeviceSurfaceFormatsKHR(device, vulkan.surface, &formatCount, nullptr);

  if (formatCount != 0) {
    vulkan.swapchainDetails.formats.resize(formatCount);
    vkGetPhysicalDeviceSurfaceFormatsKHR(device, vulkan.surface, &formatCount, vulkan.swapchainDetails.formats.data());
  }

  uint32_t presentModeCount;
  vkGetPhysicalDeviceSurfacePresentModesKHR(device, vulkan.surface, &presentModeCount, nullptr);

  if (presentModeCount != 0) {
    vulkan.swapchainDetails.presentModes.resize(presentModeCount);
    vkGetPhysicalDeviceSurfacePresentModesKHR(device, vulkan.surface, &presentModeCount, vulkan.swapchainDetails.presentModes.data());
  }
}

VkSurfaceFormatKHR chooseSwapSurfaceFormat(const std::vector<VkSurfaceFormatKHR>& availableFormats) {
  for (const auto& availableFormat : availableFormats) {
    if (availableFormat.format == VK_FORMAT_B8G8R8A8_SRGB && availableFormat.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR) {
      return availableFormat;
    }
  }

  return availableFormats[0];
}

VkPresentModeKHR chooseSwapPresentMode(const std::vector<VkPresentModeKHR>& availablePresentModes) {
  for (const auto& availablePresentMode : availablePresentModes) {
    if (availablePresentMode == VK_PRESENT_MODE_MAILBOX_KHR) {
      return availablePresentMode;
    }
  }

  return VK_PRESENT_MODE_FIFO_KHR;
}

VkExtent2D chooseSwapExtent(const VkSurfaceCapabilitiesKHR& capabilities, Vulkan& vulkan) {
  if (capabilities.currentExtent.width != UINT32_MAX) {
    return capabilities.currentExtent;
  } else {
    int width, height;
    glfwGetFramebufferSize(vulkan.window, &width, &height);

    VkExtent2D actualExtent = {
      static_cast<uint32_t>(width),
      static_cast<uint32_t>(height)
    };

    actualExtent.width = std::clamp(actualExtent.width, capabilities.minImageExtent.width, capabilities.maxImageExtent.width);
    actualExtent.height = std::clamp(actualExtent.height, capabilities.minImageExtent.height, capabilities.maxImageExtent.height);

    return actualExtent;
  }
}

void createSwapChain(Vulkan& vulkan) {
  querySwapChainSupport(vulkan.physicalDevice, vulkan);

  VkSurfaceFormatKHR surfaceFormat = chooseSwapSurfaceFormat(vulkan.swapchainDetails.formats);
  VkPresentModeKHR presentMode = chooseSwapPresentMode(vulkan.swapchainDetails.presentModes);
  VkExtent2D extent = chooseSwapExtent(vulkan.swapchainDetails.capabilities, vulkan);

  uint32_t imageCount = vulkan.swapchainDetails.capabilities.minImageCount + 2;
  if (vulkan.swapchainDetails.capabilities.maxImageCount > 0 && imageCount > vulkan.swapchainDetails.capabilities.maxImageCount)
    imageCount = vulkan.swapchainDetails.capabilities.maxImageCount;

  VkSwapchainCreateInfoKHR createInfo{};
  createInfo.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR;
  createInfo.surface = vulkan.surface;
  createInfo.minImageCount = imageCount;
  createInfo.imageFormat = surfaceFormat.format;
  createInfo.imageColorSpace = surfaceFormat.colorSpace;
  createInfo.imageExtent = extent;
  createInfo.imageArrayLayers = 1;
  createInfo.imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;

  findQueueFamilies(vulkan.physicalDevice, vulkan);
  uint32_t queueFamilyIndices[] = {vulkan.indices.graphicsFamily.value(), vulkan.indices.presentFamily.value()};

  if (vulkan.indices.graphicsFamily != vulkan.indices.presentFamily) {
    createInfo.imageSharingMode = VK_SHARING_MODE_CONCURRENT;
    createInfo.queueFamilyIndexCount = 2;
    createInfo.pQueueFamilyIndices = queueFamilyIndices;
  } else {
    createInfo.imageSharingMode = VK_SHARING_MODE_EXCLUSIVE;
    createInfo.queueFamilyIndexCount = 0;
    createInfo.pQueueFamilyIndices = nullptr;
  }

  createInfo.preTransform = vulkan.swapchainDetails.capabilities.currentTransform;

  createInfo.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;

  createInfo.presentMode = presentMode;
  createInfo.clipped = VK_TRUE;

  createInfo.oldSwapchain = VK_NULL_HANDLE;
  if (vkCreateSwapchainKHR(vulkan.device, &createInfo, nullptr, &vulkan.swapChain) != VK_SUCCESS) {
    std::cout << "Error: Failed to create swap chain" << std::endl;
    exit(EXIT_FAILURE);
  }

}

File: src/vulkan/swap_chain.rs
extern crate vulkano;

use vulkano::swapchain::{
    Surface,
    Capabilities,
    ColorSpace,
    SupportedPresentModes,
    PresentMode,
    Swapchain,
    CompositeAlpha,
};

use vulkano::format::Format;
use vulkano::image::{ImageUsage, swapchain::SwapchainImage};
use vulkano::sync::SharingMode;

use vulkano::instance::Instance;
use vulkano::device::physical::PhysicalDevice;
use vulkano::device::Device;
use vulkano::device::Queue;


use winit::window::Window;


use std::sync::Arc;


use crate::vulkan::queue_family::find_queue_families;


pub struct VkSwapChain {
    swap_chain: Option<Arc<Swapchain<Window>>>,
    swap_chain_images: Option<Vec<Arc<SwapchainImage<Window>>>>,
}

impl VkSwapChain {
    pub fn new() -> Self {
        Self {
            swap_chain: None,
            swap_chain_images: None,
        }
    }

    pub fn init(
        &mut self,
        instance: &Arc<Instance>,
        surface: &Arc<Surface<Window>>,
        physical_device_index: &usize,
        device: &Arc<Device>,
        graphics_queue: &Arc<Queue>,
        present_queue: &Arc<Queue>,
        width: &i32,
        height: &i32,
        ) {
        let physical_device = PhysicalDevice::from_index(&instance, *physical_device_index).unwrap();
        let capabilities = surface.capabilities(physical_device)
            .expect("failed to get surface capabilities");

        let surface_format = Self::choose_swap_surface_format(&capabilities.supported_formats);
        let present_mode = Self::choose_swap_present_mode(capabilities.present_modes);
        let extent = Self::choose_swap_extent(&capabilities, width, height);

        let mut image_count = capabilities.min_image_count + 1;
        if capabilities.max_image_count.is_some() && image_count > capabilities.max_image_count.unwrap() {
            image_count = capabilities.max_image_count.unwrap();
        }

        let image_usage = ImageUsage {
            color_attachment: true,
            .. ImageUsage::none()
        };

        let indices = find_queue_families(&surface, &physical_device);

        let sharing: SharingMode = if indices.graphics_family != indices.present_family {
            vec![graphics_queue, present_queue].as_slice().into()
        } else {
            graphics_queue.into()
        };

        /*
           let (swap_chain, images) = Swapchain::start(
           device.clone(),
           surface.clone(),
           image_count,
           surface_format.0,
           extent,
           1,
           image_usage,
           sharing,
           capabilities.current_transform,
           CompositeAlpha::Opaque,
           present_mode,
           true,
           None,
           ).expect("failed to create swap chain!");
           */

        let (mut swapchain, images) = {
            let caps = surface.capabilities(physical_device).unwrap();
            let composite_alpha = caps.supported_composite_alpha.iter().next().unwrap();
            let format = caps.supported_formats[0].0;
            let dimensions: [u32; 2] = surface.window().inner_size().into();

            Swapchain::start(device.clone(), surface.clone())
                .num_images(caps.min_image_count)
                .format(format)
                .dimensions(dimensions)
                .usage(ImageUsage::color_attachment())
                .sharing_mode(&queue)
                .composite_alpha(composite_alpha)
                .build()
                .unwrap()
        };

        self.swap_chain = swap_chain;
        self.swap_chain_images = images;
    }

    fn choose_swap_surface_format(available_formats: &[(Format, ColorSpace)]) -> (Format, ColorSpace) {
        *available_formats.iter()
            .find(|(format, color_space)|
                  *format == Format::B8G8R8A8Unorm && *color_space == ColorSpace::SrgbNonLinear
                 )
            .unwrap_or_else(|| &available_formats[0])
    }

    fn choose_swap_present_mode(available_present_modes: SupportedPresentModes) -> PresentMode {
        if available_present_modes.mailbox {
            PresentMode::Mailbox
        } else if available_present_modes.immediate {
            PresentMode::Immediate
        } else {
            PresentMode::Fifo
        }
    }

    fn choose_swap_extent(capabilities: &Capabilities, width: &i32, height: &i32) -> [u32; 2] {
        if let Some(current_extent) = capabilities.current_extent {
            return current_extent
        } else {
            let mut actual_extent = [width, height];
            actual_extent[0] = capabilities.min_image_extent[0]
                .max(capabilities.max_image_extent[0].min(actual_extent[0]));
            actual_extent[1] = capabilities.min_image_extent[1]
                .max(capabilities.max_image_extent[1].min(actual_extent[1]));
            actual_extent
        }
    }

    pub fn get_swap_chain(&self) -> &Arc<Surface<Window>> {
        self.swap_chain.as_ref().unwrap()
    }
}
